exporter/exporterhelper/batch_sender.go

```diff
@@ -0,0 +1,327 @@
+type BatchMergeSplitFunc func(ctx context.Context, optionalReq Request, req Request, maxItems int) ([]Request, error)
+
+type BatcherOption func(*batchSender)
+
+func WithSplitBatcher(cfg SplitBatcherConfig, msf BatchMergeSplitFunc) BatcherOption {
+	return func(b *batchSender) {
+		if cfg.MaxSizeItems != 0 {
+			b.splitCfg = cfg
+			b.mergeSplitFunc = msf
+		}
+	}
+}
+
+// batchSender is a component that accepts places requests into batches before passing them to the downstream senders.
+//
+// batch_processor implements consumer.Traces and consumer.Metrics
+//
+// Batches are sent out with any of the following conditions:
+// - batch size reaches cfg.SendBatchSize
+// - cfg.Timeout is elapsed since the timestamp when the previous batch was sent out.
+type batchSender struct {
+	baseRequestSender
+	mergeCfg       MergeBatcherConfig
+	splitCfg       SplitBatcherConfig
+	mergeFunc      BatchMergeFunc
+	mergeSplitFunc BatchMergeSplitFunc
+
+	// concurrencyLimit is the maximum number of goroutines that can be created by the batcher.
+	// If this number is reached and all the goroutines are busy, the batch will be sent right away.
+	// Populated from the number of queue consumers if queue is enabled.
+	concurrencyLimit uint64
+	activeRequests   atomic.Uint64
+
+	resetTimerCh chan struct{}
+
+	mu          sync.Mutex
+	activeBatch *batch
+
+	logger *zap.Logger
+
+	shutdownCh chan struct{}
+	stopped    *atomic.Bool
+}
+
+// newBatchSender returns a new batch consumer component.
+func newBatchSender(cfg MergeBatcherConfig, set exporter.CreateSettings, mf BatchMergeFunc, opts ...BatcherOption) requestSender {
+	bs := &batchSender{
+		activeBatch:  newEmptyBatch(),
+		mergeCfg:     cfg,
+		mergeFunc:    mf,
+		logger:       set.Logger,
+		shutdownCh:   make(chan struct{}),
+		stopped:      &atomic.Bool{},
+		resetTimerCh: make(chan struct{}),
+	}
+
+	for _, op := range opts {
+		op(bs)
+	}
+	return bs
+}
+
+func (bs *batchSender) Start(_ context.Context, _ component.Host) error {
+	timer := time.NewTimer(bs.mergeCfg.Timeout)
+	go func() {
+		for {
+			select {
+			case <-bs.shutdownCh:
+				bs.mu.Lock()
+				if bs.activeBatch.request != nil {
+					bs.exportActiveBatch()
+				}
+				bs.mu.Unlock()
+				return
+			case <-timer.C:
+				bs.mu.Lock()
+				if bs.activeBatch.request != nil {
+					bs.exportActiveBatch()
+				}
+				bs.mu.Unlock()
+				timer.Reset(bs.mergeCfg.Timeout)
+			case <-bs.resetTimerCh:
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(bs.mergeCfg.Timeout)
+			}
+		}
+	}()
+
+	return nil
+}
+
+type batch struct {
+	ctx     context.Context
+	request Request
+	done    chan struct{}
+	err     error
+}
+
+func newEmptyBatch() *batch {
+	return &batch{
+		ctx:  context.Background(),
+		done: make(chan struct{}),
+	}
+}
+
+// exportActiveBatch exports the active batch and creates a new one.
+// Caller must hold the lock.
+func (bs *batchSender) exportActiveBatch() {
+	go func(b *batch) {
+		b.err = b.request.Export(b.ctx)
```

Comment:
Sorry for the delayed response. I'm not sure I fully understand your concern. Without batching, if the timeout is not configured, a request will be blocked until the request is successfully sent or canceled. With the enabled batching, it's the same behavior: all incoming requests are blocked until the batched request is successfully sent or any of the initial requests are canceled. The latter is not fully implemented. I'll take care of it, but I believe that resolves [your previous comment](https://github.com/open-telemetry/opentelemetry-collector/pull/8685#discussion_r1378642907), not this one. Also, there is no leak, we wait until all batched requests are complete on shutdown. Please elaborate, if I misunderstood.

Reply:
> all incoming requests are blocked until the batched request is successfully sent or any of the initial requests are canceled. The latter is not fully implemented.

@dmitryax, I'm troubleshooting a customer issue where the sending queue is full and the collector will not shut down. Best I can tell it may be hanging here because it is retrying requests indefinitely. I found this comment and also noticed that https://github.com/open-telemetry/opentelemetry-collector/pull/9738 added functionality which breaks linearity between the initial requests. I don't understand the code deeply enough yet but my intuition is that context cancelation is not propagating correctly. Is that still unimplemented?